mathjax3-ts/input/tex/ParseUtil.ts

```diff
@@ -488,6 +488,123 @@ namespace ParseUtil {
 
+
+  /**
+   * Splits a package option list of the form [x=y,z=1] into an attribute list
+   * of the form {x: y, z: 1}.
+   * @param {string} attrib The attributes of the package.
+   * @param {{[key: string]: number}?} allowed A list of allowed options. If
+   *     given only allowed arguments are returned.
+   * @param {boolean?} error If true, raises an exception if not allowed options
+   *     are found.
+   * @return {EnvList} The attribute list.
+   */
+  export function keyvalOptions(attrib: string,
+                                allowed: {[key: string]: number} = null,
+                                error: boolean = false): EnvList {
+    let def: EnvList = readKeyval(attrib);
+    if (allowed) {
+      for (let key of Object.keys(def)) {
+        if (!allowed.hasOwnProperty(key)) {
+          if (error) {
+            throw new TexError('InvalidOption',
+                               'Invalid optional argument: %1', key);
+          }
+          delete def[key];
+        }
+      }
+    }
+    return def;
+  }
+
+
+  /**
+   * Implementation of the keyval function from https://www.ctan.org/pkg/keyval
+   * @param {string} text The optional parameter string for a package or
+   *     command.
+   * @return {EnvList} Set of options as key/value pairs.
+   */
+  function readKeyval(text: string): EnvList {
+    let options: EnvList = {};
+    let rest = text;
+    let end, key, val;
+    while (rest) {
+      [key, end, rest] = readValue(rest, ['=', ',']);
+      if (end === '=') {
+        [val, end, rest] = readValue(rest, [',']);
+        val = (val === 'false' || val === 'true') ?
+            JSON.parse(val) : val;
+        options[key] = val;
+      } else if (key) {
+        options[key] = true;
+      }
+    }
+    return options;
+  }
+
+
+  /**
+   * Removes pairs of outer braces.
+   * @param {string} text The string to clean.
+   * @param {number} count The number of outer braces to slice off.
+   * @return {string} The cleaned string.
+   */
+  function removeBraces(text: string, count: number): string {
+    while (count) {
+      text = text.trim().slice(1, -1);
+      count--;
+    }
+    return text.trim();
+  }
+
+
+  /**
+   * Read a value from the given string until an end parameter is reached or
+   * string is exhausted.
+   * @param {string} text The string to process.
+   * @param {string[]} end List of possible end characters.
+   * @return {[string, string, string]} The collected value, the actual end
+   *     character, and the rest of the string still to parse.
+   */
+  function readValue(text: string, end: string[]): [string, string, string] {
+    let length = text.length;
+    let braces = 0;
+    let value = '';
+    let index = 0;
+    let start = 0;
+    let startCount = true;
+    while (index < length) {
+      let c = text[index++];
+      switch (c) {
+      case '{':
+        if (startCount) {
+          start++;
+        } else if (start > braces) {
+          start = braces;
+        }
```

Comment:
There is still a small problem with this function.  It will return `}` when the input is `{{}x}` because `start` will be 2 so two pairs of terminal characters are removed, leaving only the `}`.  That is because the `}` does end up setting `startCount` to `false`, but `start is never reduced to compensate for the `}` (since the `{` is never run again).

Instead, I think you want

```
      case ' ':
        break;
      case '{':
        braces++;
        if (startCount) {
          start = braces;
        }
        break;
      case '}':
        if (braces) {
          braces--;
        }
        if (startCount) {
            start = braces;
        }
      default:
        startCount = false;
      }
```

for the cases.  This make `start` be based on the brace count when `startCount` is turned off.  (I also like that it is a bit more symmetric.)

Reply:
And I thought I had caught all corner cases... Thanks! Will correct and merge.

Reply:
The problem I see with your solution is that you again only get the count for the balanced outer braces. Thus we get the same result as my initial, incorrect solutions.
But I feel we are nearly there.

Reply:
Ah, right.  Sorry about that.  Perhaps something that include the `if-then` in the `}` case, but keeps the `start > braces` check in the `{` case?

Reply:
I think I have a solution now. It's a bit ugly, though.

Reply:
It uses two flags now: `startCount` to note that we are counting left outer brackets. `stopCount` to note that we are counting initial closing brackets. `startCount` flips on the first non-left-brace. `stopCount` is set after the first `}` and unset again after the first non-`}` after the initial `}`.
(If that makes sense.)

Reply:
I'll intersperse comments into the code to explain my thinking.

Also, this is a typical automata theory problem. I'll run it by some colleagues one of these days. There must be a more elegant solution!

Reply:
Still one case missing: `{}` is not removed.

Reply:
I think if you just include

```
if (start > braces) {
  start = braces
}
```

in the `default` case of your original version, that would do it.  (I haven't looked at your new version yet).

Reply:
Also, you would need to add

```
startCount = false;
break;
```
after the '}' case, since you don't want it to fall through to `default` after this change.

Reply:
Thanks, but I think I got it now. Should at least work for now for all non-pathological cases. 
Unless you want to have another look, I'll merge.